### ISSUE TITLE
Q: CachingUserService and RegistrationMode - is the cache used?

### ISSUE THREAD

**Issue by provegard** [IdentityServer/IdentityServer3#2075]
The default registration mode for dependencies is `RegistrationMode.InstancePerUse`. This is also used for `CachingUserService` (configured via `factory.ConfigureUserServiceCache`).

Doesn't using an instance-per-use mode defeat the purpose of having a cache?

**Comment by leastprivilege** [comment]
https://github.com/IdentityServer/IdentityServer3/blob/master/source%2FCore%2FExtensions%2FIdentityServerServiceFactoryExtensions.cs#L135

**Comment by provegard** [comment]
That doesn't answer my question, though. The relevant line is rather:

https://github.com/IdentityServer/IdentityServer3/blob/master/source%2FCore%2FExtensions%2FIdentityServerServiceFactoryExtensions.cs#L164

However, the RegistrationMode value of the Registration being set is InstancePerUse, and it's not possible to use Singleton for a factory-based Registration (https://github.com/IdentityServer/IdentityServer3/blob/a6e2b6bdfd1516af6cd9fdc9f46a08bb23bb553e/source/Core/Configuration/Hosting/AutoFacConfig.cs#L321).

**Comment by leastprivilege** [comment]
this creates a singleton instance of a client cache

https://github.com/IdentityServer/IdentityServer3/blob/master/source%2FCore%2FExtensions%2FIdentityServerServiceFactoryExtensions.cs#L138

debug it.

**Comment by provegard** [comment]
Right, but you post pointers to code that has to do with ClientStore, and the question is about UserService.

The reason I ask is that I wrote a test for my code that creates an IdentityServerOptions object, and I expected UserService to be a singleton because I configured it to be a singleton, but the test failed because its registration mode was InstancePerUse. And the reason for that is that the code calls ConfigureUserServiceCache after I configure my user service to be a singleton.

Something like this:

```
factory.UserService = new Registration<IUserService, MyUserServiceType> { Mode = RegistrationMode.Singleton };
factory.ConfigureUserServiceCache(TimeSpan.FromMinutes(5));
```

After this, the following assert fails:

```
Assert.That(factory.UserService.Mode, Is.EqualTo(RegistrationMode.Singleton));
```

With:

```
Expected: Singleton
But was:  InstancePerUse
```
